src/v/archival/upload_housekeeping_service.cc

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/upload_housekeeping_service.h"
+
+#include "archival/fwd.h"
+#include "archival/logger.h"
+#include "archival/types.h"
+#include "config/configuration.h"
+#include "ssx/future-util.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/lowres_clock.hh>
+
+#include <variant>
+
+namespace archival {
+
+std::ostream& operator<<(std::ostream& o, housekeeping_state s) {
+    switch (s) {
+    case housekeeping_state::idle:
+        o << "idle";
+        break;
+    case housekeeping_state::active:
+        o << "active";
+        break;
+    case housekeeping_state::pause:
+        o << "pause";
+        break;
+    case housekeeping_state::draining:
+        o << "draining";
+        break;
+    case housekeeping_state::stopped:
+        o << "stopped";
+        break;
+    };
+    return o;
+}
+
+upload_housekeeping_service::upload_housekeeping_service(
```

Comment:
I'm interested in the overarching design of this, having a single loop that goes through available tasks and executing them. Long term I think it can be helpful to sort/prioritize work fairly based on the current needs of the system.

That said, was this out of necessity? Could adjacent segment merging have just been added to the existing archiver housekeeping path? Curious why the architectural shift?